File: src/features/journal/types/journal.ts
export type Direction = "LONG" | "SHORT";
export type Session = "LONDON" | "NEW_YORK" | "ASIAN" | "OTHER";
export type TradeStatus = "OPEN" | "CLOSED";
export type EmotionalState =
  | "CALM"
  | "NEUTRAL"
  | "ANXIOUS"
  | "EXCITED"
  | "FEARFUL"
  | "FRUSTRATED"
  | "GREEDY";

export interface Trade extends TradeOrcaSnapshot {
  id: number;
  user_id: number;
  market: string;
  direction: Direction;
  entry_price: string | null;
  stop_loss: string | null;
  take_profit: string | null;
  exit_price: string | null;
  lot_size: string | null;
  pnl: string | null;
  rr: string | null;
  risk_pct: string | null;
  session: Session | null;
  status: TradeStatus;
  confidence: number | null;
  emotional_state: EmotionalState | null;
  stress_level: number | null;
  notes: string | null;
  trade_date: string; // ISO date string "YYYY-MM-DD"
  created_at: string;
  updated_at: string;
}

export interface TradesPage {
  trades: Trade[];
  total: number;
  page: number;
  page_size: number;
}

export interface JournalStats {
  total_trades: number;
  closed_trades: number;
  open_trades: number;
  winning_trades: number;
  losing_trades: number;
  win_rate: number;
  avg_rr: number | null;
  profit_factor: number | null;
  total_pnl: number;
  avg_pnl: number;
  best_trade_pnl: number | null;
  worst_trade_pnl: number | null;
  avg_confidence: number | null;
  avg_stress: number | null;
}

export interface CoachingSettings {
  journal_coaching_access: boolean;
}

export interface TradeCreatePayload {
  market: string;
  direction: Direction;
  entry_price?: number | null;
  stop_loss?: number | null;
  take_profit?: number | null;
  exit_price?: number | null;
  lot_size?: number | null;
  pnl?: number | null;
  rr?: number | null;
  risk_pct?: number | null;
  session?: Session | null;
  status: TradeStatus;
  confidence?: number | null;
  emotional_state?: EmotionalState | null;
  stress_level?: number | null;
  notes?: string | null;
  trade_date: string;
}

export type TradeUpdatePayload = Partial<TradeCreatePayload>;

export interface BulkImportResult {
  imported: number;
  skipped: number;
  errors: string[];
}

// ── Orca snapshot on each trade ───────────────────────────────────────────────
export interface TradeOrcaSnapshot {
  orca_score: number | null;
  orca_status: "ON" | "WATCH" | "OFF" | null;
  orca_direction: string | null;
  market_phase: string | null;
  adx_at_entry: number | null;
  ema_aligned_at_entry: boolean | null;
  intraday_bull_at_entry: number | null;
  daily_bull_at_entry: number | null;
  longterm_bull_at_entry: number | null;
  vol_score_at_entry: number | null;
}

// ── Orca analytics ────────────────────────────────────────────────────────────
export interface OrcaScoreBucket {
  label: string;
  range_min: number;
  range_max: number;
  trades: number;
  win_rate: number | null;
}

export interface OrcaStatusStat {
  status: string;
  trades: number;
  win_rate: number | null;
}

export interface OrcaPhaseStat {
  phase: string;
  trades: number;
  win_rate: number | null;
}

export interface EquityPoint {
  date: string;
  daily_pnl: number;
  cumulative_pnl: number;
}

export interface OrcaAnalytics {
  has_data: boolean;
  by_score: OrcaScoreBucket[];
  by_status: OrcaStatusStat[];
  by_phase: OrcaPhaseStat[];
}

File: backend/app/schemas/journal.py
from __future__ import annotations

from datetime import date, datetime
from decimal import Decimal
from typing import Literal, Optional

from pydantic import BaseModel, Field


Direction = Literal["LONG", "SHORT"]
Session = Literal["LONDON", "NEW_YORK", "ASIAN", "OTHER"]
TradeStatus = Literal["OPEN", "CLOSED"]
EmotionalState = Literal["CALM", "NEUTRAL", "ANXIOUS", "EXCITED", "FEARFUL", "FRUSTRATED", "GREEDY"]


class TradeCreate(BaseModel):
    market: str = Field(..., min_length=1, max_length=150)
    direction: Direction
    entry_price: Optional[Decimal] = None
    stop_loss: Optional[Decimal] = None
    take_profit: Optional[Decimal] = None
    exit_price: Optional[Decimal] = None
    lot_size: Optional[Decimal] = None
    pnl: Optional[Decimal] = None
    rr: Optional[Decimal] = None
    risk_pct: Optional[Decimal] = None
    session: Optional[Session] = None
    status: TradeStatus = "CLOSED"
    confidence: Optional[int] = Field(None, ge=1, le=10)
    emotional_state: Optional[EmotionalState] = None
    stress_level: Optional[int] = Field(None, ge=1, le=10)
    notes: Optional[str] = None
    trade_date: date


class TradeUpdate(BaseModel):
    market: Optional[str] = Field(None, min_length=1, max_length=150)
    direction: Optional[Direction] = None
    entry_price: Optional[Decimal] = None
    stop_loss: Optional[Decimal] = None
    take_profit: Optional[Decimal] = None
    exit_price: Optional[Decimal] = None
    lot_size: Optional[Decimal] = None
    pnl: Optional[Decimal] = None
    rr: Optional[Decimal] = None
    risk_pct: Optional[Decimal] = None
    session: Optional[Session] = None
    status: Optional[TradeStatus] = None
    confidence: Optional[int] = Field(None, ge=1, le=10)
    emotional_state: Optional[EmotionalState] = None
    stress_level: Optional[int] = Field(None, ge=1, le=10)
    notes: Optional[str] = None
    trade_date: Optional[date] = None


class TradeOut(BaseModel):
    id: int
    user_id: int
    market: str
    direction: Direction
    entry_price: Optional[Decimal] = None
    stop_loss: Optional[Decimal] = None
    take_profit: Optional[Decimal] = None
    exit_price: Optional[Decimal] = None
    lot_size: Optional[Decimal] = None
    pnl: Optional[Decimal] = None
    rr: Optional[Decimal] = None
    risk_pct: Optional[Decimal] = None
    session: Optional[Session] = None
    status: TradeStatus
    confidence: Optional[int] = None
    emotional_state: Optional[EmotionalState] = None
    stress_level: Optional[int] = None
    notes: Optional[str] = None
    trade_date: date
    created_at: datetime
    updated_at: datetime

    # OrcaScreener snapshot
    orca_score: Optional[int] = None
    orca_status: Optional[str] = None
    orca_direction: Optional[str] = None
    market_phase: Optional[str] = None
    adx_at_entry: Optional[int] = None
    ema_aligned_at_entry: Optional[bool] = None
    intraday_bull_at_entry: Optional[int] = None
    daily_bull_at_entry: Optional[int] = None
    longterm_bull_at_entry: Optional[int] = None
    vol_score_at_entry: Optional[int] = None

    model_config = {"from_attributes": True}


class TradesPage(BaseModel):
    trades: list[TradeOut]
    total: int
    page: int
    page_size: int


class JournalStats(BaseModel):
    total_trades: int
    closed_trades: int
    open_trades: int
    winning_trades: int
    losing_trades: int
    win_rate: float           # 0.0–100.0
    avg_rr: Optional[float]   # None if no RR data
    profit_factor: Optional[float]  # None if no losses
    total_pnl: float
    avg_pnl: float
    best_trade_pnl: Optional[float]
    worst_trade_pnl: Optional[float]
    avg_confidence: Optional[float]
    avg_stress: Optional[float]


class CoachingSettings(BaseModel):
    journal_coaching_access: bool


class EquityPointOut(BaseModel):
    date: str
    daily_pnl: float
    cumulative_pnl: float


class BulkImportResult(BaseModel):
    imported: int
    skipped: int
    errors: list[str]


class OrcaScoreBucket(BaseModel):
    label: str          # e.g. "Strong (70-100)"
    range_min: int
    range_max: int
    trades: int
    win_rate: Optional[float]


class OrcaStatusStat(BaseModel):
    status: str         # ON | WATCH | OFF | Unknown
    trades: int
    win_rate: Optional[float]


class OrcaPhaseStat(BaseModel):
    phase: str
    trades: int
    win_rate: Optional[float]


class OrcaAnalytics(BaseModel):
    has_data: bool
    by_score: list[OrcaScoreBucket]
    by_status: list[OrcaStatusStat]
    by_phase: list[OrcaPhaseStat]
